components/style/properties/properties.mako.rs

```diff
@@ -983,12 +1006,50 @@ pub struct UnparsedValue {
 }
 
 impl UnparsedValue {
-    fn substitute_variables(
-        &self,
+    fn substitute_variables<'a>(
+        &'a self,
         longhand_id: LonghandId,
         custom_properties: Option<<&Arc<::custom_properties::CustomPropertiesMap>>,
         quirks_mode: QuirksMode,
+        cache: Option<<&mut ParsedShorthandCache<'a>>,
     ) -> PropertyDeclaration {
+        match self.from_shorthand {
+            None => {
+                self.substitute_variables_from_longhand(
+                    longhand_id,
+                    custom_properties,
+                    quirks_mode,
+                )
+            }
+            Some(_) => {
+                self.substitute_variables_from_shorthand(
+                    longhand_id,
+                    custom_properties,
+                    quirks_mode,
+                    cache.expect("Must provide cache for unparsed shorthand"),
```

Comment:
This should be documented (or the no-cache case be handled, or something). It's totally unexpected that passing none in a `cache` argument will suddenly crash your program.

Reply:
Handling no-cache case could be a bit tricky. I tried to put a local cache in `substitute_variables_from_shorthand` and use it when cache is not provided, but given this struct is big, the stack allocated on stack is probably non-trivial, so I decided not to do that, but instead just require callsites to provide it when necessary.

I agree that there should be some document.